<a href="https://colab.research.google.com/github/EliMLevy/Artificial-Intelligence-COM3760/blob/main/HW4/knnHW.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Import the libraries math (for square root and absolute value) and pandas and numpy for reading csv files.

In [40]:
import math
import pandas as pd
import numpy as np

Make an untagged vector, point, and two tagged vectors, data1 and data2:

In [2]:
point = [1, 0, 0, '?'] #(an unknown tag)
data1 = [1, 1, 1, 'M']
data2 = [1, 3, 0, 'F']


Write code to separate the data (X) from the tag (Y).  Your output should be:

The vector [1, 1, 1] has tag M

The vector [1, 3, 0] has tag F


In [5]:

def separate_data_from_tag(data):
  return data[0:-1], data[-1]

data, tag = separate_data_from_tag(data1)
print("The vector " , data, " has tag ", tag)
data, tag = separate_data_from_tag(data2)
print("The vector " , data, " has tag ", tag)

The vector  [1, 1, 1]  has tag  M
The vector  [1, 3, 0]  has tag  F


Now let's classify the point as either M or F.  We'll do this by setting k = 1 and using the Euclidean distance.  We'll define that as:

In [7]:
def euclideanDistance(instance1, instance2, length):
   distance = 0
   for x in range(length):
         #print ('x is ' , x)
         num1=float(instance1[x])
         num2=float(instance2[x])
         distance = distance + pow(num1-num2, 2)
   return math.sqrt(distance)


Now, find out the distance between data1 and point and data2 and point. Output what the tag should be using euclideanDistance.

In [12]:

data1_dist = euclideanDistance(data1,point,3)
data2_dist = euclideanDistance(data2,point,3)

print("data1 dist: ", data1_dist)
print("data2 dist: ", data2_dist)

if data1_dist < data2_dist:
  _, tag = separate_data_from_tag(data1)
else:
  _, tag = separate_data_from_tag(data2)

print("The tag should be ", tag)

data1 dist:  1.4142135623730951
data2 dist:  3.0
The tag should be  M


Now, let's get more data from a file, myFile.csv. This code will read it for us:

In [13]:
url = 'https://github.com/rosenfa/ai/blob/master/myFile.csv?raw=true'
df=pd.read_csv(url,  header=0, error_bad_lines=False)
#put data in dataset without header line
dataset = np.array(df)

<ipython-input-13-433b83a05c7f>:2: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  df=pd.read_csv(url,  header=0, error_bad_lines=False)


Show that you understand what you read by:

1. Printing the first two vectors in the file
2. Printing the Euclidean distance between those two vectors

Here is some code which I think might help get you on your way!

In [63]:
print(len(dataset))
# print(dataset)
# print(euclideanDistance(dataset[0],dataset[1],3))

print(dataset[:2])
print(euclideanDistance(dataset[0], dataset[1], 3))

10
[[0 1 2 'F']
 [1 5 6 'F']]
5.744562646538029


Now assume you have a new value for point:

[0,1,100]

How would you classify this vector using the Euclidean distance function given all of the vectors in the file?

In order to help here is a hint:  

We suggest defining some type of data struction to store different vectors' distances and their tags like this:

In [22]:
point=[0,1,100]
class distClass:
    dist = -1 #distance of current point from test point
    tag = '-' #tag of current point


You can then add vector distances like this:

In [23]:
eucDistances = [] # list of distances, will hold objects of type distClass

temp=dataset[1]
label=temp[-1]
d=euclideanDistance(point,temp,3)
print("The distances between " , point , " and " , temp,  " is " , str(d))
print(" and the label is " + label)
obj = distClass() #one record's distance and tag
obj.dist=d
obj.tag=label
eucDistances.append(obj)


The distances between  [0, 1, 100]  and  [1 5 6 'F']  is  94.09038208021052
 and the label is F


In [64]:
# Calculates the distance between the point and the specified row using the given distance function
def process_row(point, row, container, length, dist_func):
  label=row[-1]
  d=dist_func(point,row,length)
  # print("The distances between " , point , " and " , temp,  " is " , str(d))
  # print(" and the label is " + label)
  obj = distClass() #one record's distance and tag
  obj.dist=d
  obj.tag=label
  container.append(obj)

In [ ]:
eucDistances[0].dist

94.13819628609845

and sort them like this:

In [25]:
eucDistances.sort(key=lambda x: x.dist)

Questions:

1. What is the label for point if k=1?
2. What is the label for point if k=3?
3. Would the result be different if we used a different distance function like Hamming or Manhattan?

In [69]:
from collections import defaultdict

'''
Given a list of distClass find the label of the majority of the k nearest neighbors
'''
def categorize(lst, k):
  lst.sort(key=lambda x: x.dist)
  results = defaultdict(int)
  best_result = lst[0].tag
  for i in range(k):
    # print("looking at ", i, " with tag: ", lst[i].tag)
    results[lst[i].tag] += 1
    if results[lst[i].tag] > results[best_result]:
      best_result = lst[i].tag
  # print(results)
  return best_result


In [67]:
'''
Finds the knn of the specified point with the euclidean distance method
'''
distances = []
for row in dataset:
  process_row(point, row, distances, 3, euclideanDistance)


k1_category = categorize(distances, 1)
k3_category = categorize(distances, 3)
k7_category = categorize(distances, 7)
print("With k=1, category is ", k1_category)
print("With k=3, category is ", k3_category)
print("With k=7, category is ", k7_category)

defaultdict(<class 'int'>, {'F': 1})
defaultdict(<class 'int'>, {'F': 2, 'M': 1})
defaultdict(<class 'int'>, {'F': 4, 'M': 3})
With k=1, category is  F
With k=3, category is  F
With k=7, category is  F


In [41]:
#Add code with functions that implement Hamming and Manhattan distances and test
#what the label will be for k=1 and k=3 for all possibilities
#(6 total: 2 Euclidean, 2 Hamming and 2 Manhattan)

def hammingDistance(instance1, instance2, length):
  # Add one to count if they are not the same
  count = 0
  for i in range(length):
    if instance1[i] != instance2[i]:
      count += 1

  return count

def manhattanDistance(instance1, instance2, length):
  dist = 0
  for i in range(length):
    dist += abs(instance1[i] - instance2[i])

  return dist



In [43]:
'''
Finds the knn of the specified point with the hamming distance method
'''
distances = []
for row in dataset:
  process_row(point, row, distances, 3, hammingDistance)


k1_category = categorize(distances, 1)
k3_category = categorize(distances, 3)
print("Hamming: With k=1, category is ", k1_category)
print("Hamming: With k=3, category is ", k3_category)

Hamming: With k=1, category is  F
Hamming: With k=3, category is  F


In [44]:
'''
Finds the knn of the specified point with the manhattan distance method
'''
distances = []
for row in dataset:
  process_row(point, row, distances, 3, manhattanDistance)


k1_category = categorize(distances, 1)
k3_category = categorize(distances, 3)
print("Manhattan: With k=1, category is ", k1_category)
print("Manhattan: With k=3, category is ", k3_category)

Manhattan: With k=1, category is  M
Manhattan: With k=3, category is  F


Up until here is a simplified version of the homework.
Below here is the work for the part we will be checking as the basis of your grade:

Now let's look at some bigger files:

1. mytrain.csv (for training the model)
2. mytest.csv (for testing)


In [45]:
url = 'https://github.com/rosenfa/ai/blob/master/mytrain.csv?raw=true'
train_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))
url = 'https://github.com/rosenfa/ai/blob/master/mytest.csv?raw=true'
test_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))

<ipython-input-45-d4827212fd82>:2: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  train_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))
<ipython-input-45-d4827212fd82>:4: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  test_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))


In [46]:
print(train_data.shape)# number of records and features
print(train_data)

(100, 30)
[[0.317251 0.540016 0.150514 ... 0.0 0.212007 'F']
 [3.882181 0.0 0.0 ... 0.0 0.0 'F']
 [1.790708 0.0 0.0 ... 0.0 0.0 'F']
 ...
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'F']]


In [47]:
print(test_data.shape)# number of records and features
print(test_data)

(100, 30)
[[2.85985 0.66275 0.0 ... 0.995289 0.0 'F']
 [0.0 12.998726 0.0 ... 0.0 0.0 'F']
 [13.231318 0.0 0.0 ... 0.0 0.0 'F']
 ...
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [1.688611 0.0 3.204529 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'M']]


I hope by now you understand where we are going with this :)

Now implement the knn code with 3 different values for k:
1. k = 1
2. k = 7
3. k = 15

and at first use the Euclidean distance.
Classify each of the vectors in the test_data dataset using the training data from train_data.  Which value for k gave the best accuracy?  What accuracy did it give you?

Now see if using Hamming or Manhattan distance give any better results for the same values of k.  

Once you are done, you should have a total of 9 different results:

1. Three accuracy results for the different value of k using the Euclidean Distance
2. Three accuracy results for the different value of k using the Hamming Distance
3. Three accuracy results for the different value of k using the Manhattan Distance

Hint: I strongly suggest you base yourself on the code you've seen until this point so you don't have to reinvent the wheel!

Accuracy = (TP + TN) / (TP + TN + FT + FN)

Accuracy = (Total correct) / (Total Guesses)

In [48]:
features = test_data.shape[1] - 1
print(features)

29


In [78]:
#Add code here

def evaluate(dist_func):
  k1_correct = 0
  k7_correct = 0
  k15_correct = 0
  for point in test_data:
    distances = []
    for row in train_data:
      process_row(point, row, distances, features, dist_func)

    k1_category = categorize(distances, 1)
    k7_category = categorize(distances, 7)
    k15_category = categorize(distances, 15)
    if k1_category == point[-1]:
      k1_correct += 1
    if k7_category == point[-1]:
      k7_correct += 1
    if k15_category == point[-1]:
      k15_correct += 1

  return k1_correct, k7_correct, k15_correct


k1_correct, k7_correct, k15_correct = evaluate(euclideanDistance)
print("Euclidean")
print("k1 correct: ", k1_correct, ". Accuracy: ", k1_correct / len(test_data))
print("k7 correct: ", k7_correct, ". Accuracy: ", k7_correct / len(test_data))
print("k15 correct: ", k15_correct, ". Accuracy: ", k15_correct / len(test_data))
print()

k1_correct, k7_correct, k15_correct = evaluate(hammingDistance)
print("Hamming")
print("k1 correct: ", k1_correct, ". Accuracy: ", k1_correct / len(test_data))
print("k7 correct: ", k7_correct, ". Accuracy: ", k7_correct / len(test_data))
print("k15 correct: ", k15_correct, ". Accuracy: ", k15_correct / len(test_data))
print()

k1_correct, k7_correct, k15_correct = evaluate(manhattanDistance)
print("Manhattan")
print("1 correct: ", k1_correct, ". Accuracy: ", k1_correct / len(test_data))
print("7 correct: ", k7_correct, ". Accuracy: ", k7_correct / len(test_data))
print("15 correct: ", k15_correct, ". Accuracy: ", k15_correct / len(test_data))


Euclidean
k1 correct:  50 . Accuracy:  0.5
k7 correct:  74 . Accuracy:  0.74
k15 correct:  70 . Accuracy:  0.7

Hamming
k1 correct:  61 . Accuracy:  0.61
k7 correct:  55 . Accuracy:  0.55
k15 correct:  57 . Accuracy:  0.57

Manhattan
1 correct:  61 . Accuracy:  0.61
7 correct:  63 . Accuracy:  0.63
15 correct:  69 . Accuracy:  0.69


The value of K that gave me the best accuracy was k=7 with the Euclidean distance. The accuracy was 74%.

Grade Basis:

80% for correct answers (and yes, there are possibilities that multiple answers are correct-- especially for cases of ties).

20% : Documentation and easily readable code

Please publish your final Notebook in your Github directory.

The homework assignment is due by November 27th at 11:59 PM
